cherab/tools/observers/bolometry.py

```diff
@@ -687,6 +695,350 @@ def etendue_single_run(_):
+            pipeline = SpectralRadiancePipeline2D(display_progress=False)
+        else:
+            raise ValueError("Sensitivity units can only be of type 'Power' or 'Radiance'.")
+
+        voxel_collection.set_active("all")
+
+        cached_max_wavelength = self.max_wavelength
+        cached_min_wavelength = self.min_wavelength
+        cached_bins = self.spectral_bins
+        cached_pipelines = self.pipelines
+        cached_ray_count = self.pixel_samples
+
+        self.pipelines = [pipeline]
+        self.min_wavelength = 1
+        self.max_wavelength = voxel_collection.count + 1
+        self.spectral_bins = voxel_collection.count
+        self.pixel_samples = ray_count
+
+        self.observe()
+
+        self.max_wavelength = cached_max_wavelength
+        self.min_wavelength = cached_min_wavelength
+        self.spectral_bins = cached_bins
+        self.pipelines = cached_pipelines
+        self.pixel_samples = cached_ray_count
+
+        return pipeline.frame.mean
+
+    def calculate_etendue(self, ray_count=10000, batches=10, max_distance=1e999):
+        """
+        Calculates the etendue of each pixel in this detector.
+
+        This function calculates the detector etendue by evaluating the ratio of rays that
+        pass un-impeded through the detector's aperture. For this method to work, the detector
+        and its aperture structures should be the only primitives present in the scene. If any
+        other primitives are present, the results may be misleading.
+
+        :param int ray_count: The number of rays used per batch.
+        :param int batches: The number of batches used to estimate the error on the etendue
+          calculation.
+        :param float max_distance: The maximum distance from the detector to consider intersections.
+            If a ray makes it further than this, it is assumed to have passed through the aperture,
+            regardless of what it hits. Use this if there are other primitives present in the scene
+            which do not form the aperture.q
+        :return: a tuple (etendue, etendue_error), each of which is a 2D
+          array of size (ncol, nrow)
+        """
+        # Calculate the etendue of each pixel as a separate task
+        nx, ny = self.pixels
+        pixels_flattened = self.pixels_as_foils.ravel()
+        etendues_flattened = np.empty_like(pixels_flattened, dtype=float)
+        etendue_errors_flattened = np.empty_like(etendues_flattened)
+
+        def calculate(pixel_index):
+            foil = pixels_flattened[pixel_index]
+            foil.render_engine.processes = 1
+            return pixel_index, foil.calculate_etendue(ray_count, batches, max_distance)
+
+        def update(result):
+            index, (etendue, etendue_error) = result
+            etendues_flattened[index] = etendue
+            etendue_errors_flattened[index] = etendue_error
+
+        self.render_engine.run(list(range(nx * ny)), calculate, update)
```

Comment:
Elegant use of nested functions to use the render engine. You have earned 1 free espresso from our office.